listenbrainz/webserver/static/js/src/recommendations/RecommendationCard.test.tsx

```diff
@@ -16,25 +22,297 @@ const recommendation: Recommendation = {
+    spy.mockImplementation(() => Promise.resolve(200));
+
+    // Check initial values of HTML elements
+    expect(
+      wrapper.find(".recommendation-controls").childAt(0).hasClass("love")
+    ).toEqual(true);
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(0).prop("icon").iconName).toEqual("grin-stars");
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(2).text()).toEqual("Love");
+    await instance.submitFeedback("hate");
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith("lalala", "yyyy", "hate");
+
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(1);
+    expect(instance.props.updateFeedback).toHaveBeenCalledWith("yyyy", "hate");
+
+    // Emulate updating the props after calling Recommendations component's updateFeedback
+    wrapper.setProps({ currentFeedback: "hate" });
+
+    // Check that HTML elements have changed accordingly
+    expect(
+      wrapper.find(".recommendation-controls").childAt(0).hasClass("hate")
+    ).toEqual(true);
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(0).prop("icon").iconName).toEqual("angry");
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(2).text()).toEqual("Hate");
+  });
+
+  it("does nothing if isCurrentUser is false", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard
+        {...{ ...props, isCurrentUser: false, updateFeedback: jest.fn() }}
+      />
+    );
+    const instance = wrapper.instance();
+
+    const spy = jest.spyOn(instance.APIService, "submitRecommendationFeedback");
+    spy.mockImplementation(() => Promise.resolve(200));
+
+    instance.submitFeedback("dislike");
+    expect(spy).toHaveBeenCalledTimes(0);
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(0);
+    expect(wrapper.exists(".recommendation-controls")).toEqual(false);
+  });
+
+  it("does nothing if CurrentUser.authtoken is not set", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard
+        {...{
+          ...props,
+          currentUser: { auth_token: undefined, name: "test" },
+          updateFeedback: jest.fn(),
+        }}
+      />
+    );
+    const instance = wrapper.instance();
+
+    const spy = jest.spyOn(instance.APIService, "submitRecommendationFeedback");
+    spy.mockImplementation(() => Promise.resolve(200));
+
+    instance.submitFeedback("love");
+    expect(spy).toHaveBeenCalledTimes(0);
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(0);
+    expect(wrapper.exists(".recommendation-controls")).toEqual(false);
+  });
+
+  it("doesn't call updateFeedback if status code is not 200", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard {...props} />
+    );
+    const instance = wrapper.instance();
+    props.updateFeedback = jest.fn();
```

Comment:
Something I hadn't caught earlier that I think negates all the tests that have this structure (`props.updateFeedback = jest.fn();` followed by `expect(props.updateFeedback)…`).

I believe it's replacing the function in the `props` object created at the top of the test file, not modifying the props of the component being tested. The wrapper or instance won't automatically update its props like that, especially since we're passing a clone of the props (with the spread syntax to copy the object `{…props}`) to start with.

Instead it should be something like 
```
    const updateFeedbackSpy = jest.spyOn(instance.props, "updateFeedback");
[...]
    expect(updateFeedbackSpy).toHaveBeenCalledTimes(0);
```

I spotted 6 cases in this test file, but it's worth having another look and making sure we don't do that anywhere else.